packages/react-sdk/src/internals/useAction.ts

```diff
@@ -0,0 +1,54 @@
+import { useCallback, useState } from 'react'
+import type { SyncAgent } from '@river-build/sdk'
+import { useSyncAgent } from '../useSyncAgent'
+
+export type ActionConfig = {
+    onError?: (err: Error) => void
+    // onSucess?: (data: T) => void
+}
+
+type MultipleParams<T> = T extends unknown[] ? T : [T]
+
+export const useAction = <Data, Params>(
+    fn: (sync: SyncAgent) => (...args: MultipleParams<Params>) => Promise<Data>,
+    config: ActionConfig = {},
+) => {
+    const [status, setStatus] = useState<'loading' | 'error' | 'success' | 'idle'>('idle')
+    const [error, setError] = useState<Error | undefined>()
+    const [data, setData] = useState<Data | undefined>()
+
+    const sync = useSyncAgent()
+
+    const action = useCallback(
+        async (...args: MultipleParams<Params>) => {
+            setStatus('loading')
+            try {
+                const data = await fn(sync)(...args)
```

Comment:
The idea of a wrapper hook is a good one, but the implementation is tricky 😬

When we pass 
```ts
useAction(sync => sync.spaces.createSpace) // <-- the action from sync agent
```

`createSpace` is a _method_ on the Spaces **class**. Which means this is using `this` everywhere.

When we do 

```ts
block: {
    await fn(sync)(...args)
    //  javascript`this` here is "block", not the originally bind class anymore.
}
```

How we fix? 
Patching the context. Inject the right `this` via [`fn.call`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/call) / [`fn.apply`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/apply)

We need a way to rebuild the `sync.spaces.createSpace` path from `createSpace` method, so we can patch it with the right `this` context (in this example, `sync.spaces`)

This could be achieved by searching all the keys in the sync class, but due to circular references this isn't ideal.

I'll pause the `useAction` internal wrapper hook for now, unless anyone come with a good solution 


this sucks and im mad at it

Reply:
Check out useWithCatch in harmony repo, I did a similar thing https://github.com/HereNotThere/harmony/blob/4f18f9865a4cfb4fce08f5de487924205fc17385/clients/web/lib/src/hooks/use-towns-client.ts#L368

Reply:
In that case, you have a flatten structure: `<class>.<set-action>`, so it's safe to just apply the class as `this`. Here we have `<class>.<namespace>.<set-action>`.

Now we need to know the namespace of the action to provide the correct `this` context, but the useAction hook only knows `<class>` and `<set-action>`. It's missing information to recreate the full path

Reply:
Aha, since it's an internal hook only, I think a possible solution would be to pass the namespace and then the action
```ts
useAction(sync.channel, "createChannel")
// first you pass the namespace, the second one will list all the possible actions in this namespace
```
This way we have the right context for the action